app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -57,58 +65,138 @@ export class PersistenceManager {
    * number of the state tree.
    */
-  #metadata?: { version: number };
+  #metadata?: MetaData;
 
   #isExtensionInitialized: boolean = false;
 
-  #localStore: ExtensionStore | ReadOnlyNetworkStore;
+  #localStore: BaseStore;
+
+  #db: DB;
+
+  #open: boolean = false;
 
-  constructor({
-    localStore,
-  }: {
-    localStore: ExtensionStore | ReadOnlyNetworkStore;
-  }) {
+  constructor({ localStore }: { localStore: BaseStore }) {
     this.#localStore = localStore;
+    this.#db = new DB();
+  }
+
+  async open() {
+    if (!this.#open) {
+      await this.#db.open('metamask-vault', 1);
+      this.#open = true;
+    }
   }
 
-  setMetadata(metadata: { version: number }) {
+  setMetadata(metadata: MetaData) {
     this.#metadata = metadata;
   }
 
+  /**
+   * Sets the state in the local store.
+   *
+   * @param state - The state to set in the local store. This should be an object
+   * containing the state data to be stored.
+   * @throws Error if the state is missing or if the metadata is not set before
+   * calling this method.
+   * @throws Error if the local store is not open.
+   * @throws Error if the data persistence fails during the write operation.
+   */
   async set(state: MetaMaskStateType) {
+    await this.open();
+
     if (!state) {
       throw new Error('MetaMask - updated state is missing');
     }
-    if (!this.#metadata) {
+    const meta = this.#metadata;
+    if (!meta) {
       throw new Error('MetaMask - metadata must be set before calling "set"');
     }
-    try {
-      await this.#localStore.set({ data: state, meta: this.#metadata });
-      if (this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = false;
-      }
-    } catch (err) {
-      if (!this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = true;
-        captureException(err);
-      }
-      log.error('error setting state in local store:', err);
-    } finally {
-      this.#isExtensionInitialized = true;
-    }
+
+    await navigator.locks.request(
+      STATE_LOCK,
+      { mode: 'exclusive' },
+      async () => {
+        try {
+          // atomically set all the keys
+          await this.#localStore.set({
+            data: state,
+            meta,
+          });
+
+          // back up the vault in IndexedDB
+          const keyringController =
+            state.KeyringController as KeyringControllerState;
+          const vault = keyringController?.vault;
+          if (vault) {
+            await this.#db.set({ vault });
+          } else {
+            // if we don't have a vault, remove the backup from IndexedDB
+            // TODO: should we create a backup for the backup when a vault is
+            // removed?
+            await this.#db.remove(['vault']);
+          }
+
+          if (this.#dataPersistenceFailing) {
+            this.#dataPersistenceFailing = false;
+          }
+        } catch (err) {
+          if (!this.#dataPersistenceFailing) {
```

Comment:
if we CAN write to the primary database (`this.#localStore.set`), but we writing to the _backup_ database files (`this.#db.set`)...  should we still treat it as `dataPersistenceFailing`? This code currently behaves this way. 🤔 

From the comment on `#dataPersistenceFailing`:

```
  /**
   * dataPersistenceFailing is a boolean that is set to true if the storage
   * system attempts to write state and the write operation fails. This is only
   * used as a way of deduplicating error reports sent to sentry as it is
   * likely that multiple writes will fail concurrently.
   */
```
   
we only use this as a flag to dedupe error events, and not for anything else.